allocator.go

```diff
@@ -155,6 +176,17 @@ func rangeTagsAreSubsetOfNodeTags(rangeTags []string, nodeTags []string) bool {
 	return true
 }
 
+func toMap(t map[RangeID][]NodeID) map[RangeID]map[NodeID]int {
+	ret := make(map[RangeID]map[NodeID]int)
```

Comment:
Instead of an int, you could use struct{}. Then to check if a key exists, you'd do `_, ok := o[r.id][n.id]`. Btw, if you provide a r.id that doesn't exist (cause it's a new range), you'll have an NPE. I think a better way to do it would be creating a private tuple type; a rangeID+nodeID pair. Something like https://github.com/irfansharif/allocator/blob/96fda6c95716d65eaf4999199abae63b06a9ed49/allocator.go#L41.

Reply:
@AlirezaDX this is a valid concern, can we have this addressed?

Reply:
Creating a private tuple would cause a loss of information. Since the map would be only indexed with the combination of rangeID+nodeID, there would be no easy way to understand whether the range is missing or the node if needed.

The nil value of a map is an empty map, therefore the usage below would not throw an NPE given an unknown range:
```GoLang
if previousAssignment, ok := prevAssignmentMap[r.id][n.id]; ok == true && previousAssignment == 1
```

Given the two above reasons I don't think doing so is necessary.

Reply:
@irfansharif let me know if I am missing the part where an NPE can occur.

Reply:
https://programming.guide/go/default-zero-value.html, wouldn't the map returned by an invalid `r.id` be `nil`?

Reply:
Yea, what Saad said. Basically it's possible for a new range ID to be passed in during a second call to Allocate, and that would NPE out. I don't understand your loss of information comment. We're not currently checking if a particular node or range exists. Besides, I think all this map stuff goes away once we start modeling replica vars as int vars instead if literals. I'll defer to you guys for what you want to do for this PR -- it'd be nice to merge soon.

Reply:
Please give my comments a chance and run some test code:
```Golang
func TestMap(t *testing.T) {
	var myMap = make(map[int]map[int]int)
	myMap[2] = map[int]int{2: 1, 3: 2}

	fmt.Println(myMap)
	fmt.Println(myMap[1][1])
	if _, ok := myMap[1][1]; ok {
		fmt.Println("Not printing this AND no NPE")
	}
}
```
Above wouldn't throw an NPE. Let me know if I am still missing the part that it can be throwing an NPE.

Reply:
Note that a `nil` map is only read as an empty map and writing to the map would actually cause an NPE.

Reply:
`A nil map is equivalent to an empty map except that no elements may be added.` From : https://golang.org/ref/spec#Map_types
Interesting, I didn't read into it deep enough.